src/main/scala/org/apache/mesos/chronos/scheduler/jobs/JobStats.scala

```diff
@@ -124,106 +166,157 @@ class JobStats @Inject()(clusterBuilder: Option[Cluster.Builder], config: Cassan
   }
 
   def jobFinished(job: BaseJob, taskStatus: TaskStatus, attempt: Int) {
-    try {
-      getSession match {
-        case Some(session: Session) =>
-          job match {
-            case job: ScheduleBasedJob =>
-              val query =
-                s"INSERT INTO ${config.cassandraTable()} (id, ts, job_name, job_owner, job_schedule, task_state, slave_id, attempt) VALUES (?, ?, ?, ?, ?, ?, ?, ?) USING TTL ${config.cassandraTtl()}"
-              val prepared = statements.getOrElseUpdate(query, {
-                session.prepare(
-                  new SimpleStatement(query).setConsistencyLevel(ConsistencyLevel.valueOf(config.cassandraConsistency())).asInstanceOf[RegularStatement]
-                )
-              })
-              session.executeAsync(prepared.bind(
-                taskStatus.getTaskId.getValue,
-                new java.util.Date(),
-                job.name,
-                job.owner,
-                job.schedule,
-                taskStatus.getState.toString,
-                taskStatus.getSlaveId.getValue,
-                attempt: java.lang.Integer
-              ))
-            case job: DependencyBasedJob =>
-              val query =
-                s"INSERT INTO ${config.cassandraTable()} (id, ts, job_name, job_owner, job_parents, task_state, slave_id, attempt) VALUES (?, ?, ?, ?, ?, ?, ?, ?) USING TTL ${config.cassandraTtl()}"
-              val prepared = statements.getOrElseUpdate(query, {
-                session.prepare(
-                  new SimpleStatement(query).setConsistencyLevel(ConsistencyLevel.valueOf(config.cassandraConsistency())).asInstanceOf[RegularStatement]
-                )
-              })
-              val parentSet: java.util.Set[String] = job.parents.asJava
-              session.execute(prepared.bind(
-                taskStatus.getTaskId.getValue,
-                new java.util.Date(),
-                job.name,
-                job.owner,
-                parentSet,
-                taskStatus.getState.toString,
-                taskStatus.getSlaveId.getValue,
-                attempt: java.lang.Integer
-              ))
-          }
-        case None =>
-      }
-    } catch {
-      case e: NoHostAvailableException =>
-        resetSession()
-        log.log(Level.WARNING, "No hosts were available, will retry next time.", e)
-      case e: QueryExecutionException =>
-        log.log(Level.WARNING, "Query execution failed:", e)
-      case e: QueryValidationException =>
-        log.log(Level.WARNING, "Query validation failed:", e)
+    updateJobState(job.name, CurrentState.idle)
+    job match {
+      case job: ScheduleBasedJob =>
+        insertToStatTable(
```

Comment:
A lot of this code is duplicated between the cases. Let's separate out the common logic.